src/price_data.py

```diff
@@ -326,3 +328,133 @@ def get_cost(self, tr: Union[Operation, SoldCoin], reference_coin: str = config.
+                if market[0] == coin and market[1] == reference_coin:
+                    return [coin, reference_coin, False]  # False=not inverted
+                elif market[1] == coin and market[0] == reference_coin:
+                    return [reference_coin, coin, True]  # True=inverted
+
+        pair = get_pair(coin, reference_coin)
+        if not pair:
+
+            for market in self.markets:
+                pair = get_pair(market[1], reference_coin)
+
+                if pair:
+                    if market[0] == coin:
+                        return [[market[0], market[1], False], pair]
+                    elif market[1] == coin:
+                        return [[market[1], market[0], True], pair]
+        else:
+            return [pair, ]
+
+    def _get_binance_bulk_pair_data(self, operations: list, symbol: str, invert: str=False) ->list:
+        timestamps = []
+        timestamppairs = []
+        data = []
+
+        for op in operations:
+            timestamps.append(op.utc_time)
+
+        while len(timestamps) > 0:
+            timestamp = timestamps.pop(0)
+
+            if len(timestamppairs) > 0 and timestamppairs[-1][0]+datetime.timedelta(minutes=995) > timestamp:
+                timestamppairs[-1].append(timestamp)
+            else:
+                timestamppairs.append([timestamp, ])
+
+        for batch in timestamppairs:
+            # ccxt works with timestamps in milliseconds
+            last = int(max(batch).timestamp() * 1000)
+            first = int(min(batch).timestamp() * 1000)
+
+            if invert:
+                tempdata = list(
+                    map(lambda x: (x[0], 1/((x[1]+x[4])/2)), self.get_candles(first, last, symbol)))
+            else:
+                tempdata = list(
+                    map(lambda x: (x[0], (x[1]+x[4])/2), self.get_candles(first, last, symbol)))
+
+            if tempdata:
+                for operation in batch:
+                    # TODO discuss which candle is picked current is closest to original date (often off by about 1-20s, but can be after the Trade)
+                    # times do not always line up perfectly so take one nearest
+                    ts = list(
+                        map(lambda x: (abs(operation.timestamp()*1000-x[0]), x), tempdata))
+                    data.append((operation, min(ts, key=lambda x: x[0])[1][1]))
+        return data
+
+    def preload_price_data(self, operations: list, coin: str):
+
+        reference_coin = config.FIAT
+        # get pairs used for calculating the price
+        lis = self._get_binance_bulk_pair_list(reference_coin, coin)
+        db_path = self.get_db_path("binance")
+        operations_filtered = []
+        tablename = self.get_tablename(coin, reference_coin)
+
+        if lis:
+
+            for operation in operations:
+                if not self.__get_price_db(db_path, tablename, operation.utc_time):
+                    operations_filtered.append(operation)
```

Comment:
1. this is a point of optimization. I currently only implemented it that way, because in most cases (except for bnb at binance) the datasets should be relatively small (under 20k for bnb and under 1k most of the times for other pairs) 
2. yes good point (man i love list comprehensions i always forget that they exist when coding)